### PR TITLE
Update request form

### PR DESCRIPTION
**Pull request by ShellyXueHan**
1. make IDIR a mandatory identity provider
2. update form existing fields
3. extent form based on required list
4. update request list filtering based on the user role
5. limit the number of open request for be 1 for every requester
6. some UI fix

### CODE REVIEW

**Comment by patricksimonian** [comment]
Due to our first-quarter profits not being as high as we hoped, pr reviews are now being done at  a cost of $5.00 per review 😄

**Comment by ShellyXueHan** [comment]
@patricksimonian Due to our first-quarter profits being negative, we are starting to charge $10.99/review for all passed reviews. Payment expected to arrive in a week, any delay could result in user access removal of the platform ;P

File: api/src/constants/form.js [hunk 1]
```diff
@@ -40,10 +40,30 @@ export const FORM_CONTENT_TO_REQUEST = {
...
+const BCEID_CONTACT_OBJECT = role => {
+  return {
+    name: `bceid.contactInfo.${role}.name`,
+    title: `bceid.contactInfo.${role}.title`,
+    email: `bceid.contactInfo.${role}.email`,
+  };
+};
+
+// Map role to each contact object:
+export const BCEID_CONTACT_INFO_SCHEMA = roles => {
```
Comment (patricksimonian) [inline api/src/constants/form.js]: naming function as a constant seems interesting 🤔

Comment (patricksimonian) [reply api/src/constants/form.js]: I'd call it mapRolesToContact or something similar?

File: api/src/constants/form.js [hunk 2]
```diff
@@ -60,6 +80,10 @@ export const REQUEST_TO_FORM_CONTENT = {
   forecastAmount: 'bceid.forecastAmount',
   prodDate: 'bceid.prodDate',
   useDate: 'bceid.useDate',
+  // v2:
+  appName: 'bceid.appName',
+  contactInfo: BCEID_CONTACT_INFO_SCHEMA(['es', 'pm', 'tl', 'pl', 'sl', 'cl']),
```
Comment (patricksimonian) [inline api/src/constants/form.js]: are those values constant? (es, pm, tl, pl, sl, cl) what do they mean? where else can they be found?

File: api/src/router/routes/github-webhook.js [hunk 3]
```diff
@@ -101,6 +101,7 @@ router.post(
           break;
         case PR_ACTIONS.LABELED:
           // Obtain the labels assigned to the PR
+          // eslint-disable-next-line no-case-declarations
```
Comment (patricksimonian) [inline api/src/router/routes/github-webhook.js]: why is this needed?

Comment (patricksimonian) [reply api/src/router/routes/github-webhook.js]: the nested switch is quite large, for cleanliness it could be broken up into its own function (just a thought !)

File: api/src/router/routes/github.js [hunk 4]
```diff
@@ -45,11 +45,14 @@ router.put(
     const { approvalContent } = req.body;
     try {
       // TODO: handle rejection message
+      // eslint-disable-next-line no-unused-vars
       const { isApproved, message } = approvalContent;
 
       // Alter between labels of rejected and ready:
-      if (!isApproved) await alterPRLabels(prNumber, GITHUB_LABELS.BCEID_APPROVED, GITHUB_LABELS.BCEID_REJECTED);
-      else await alterPRLabels(prNumber, GITHUB_LABELS.BCEID_REJECTED, GITHUB_LABELS.BCEID_APPROVED);
+      if (!isApproved)
```
Comment (patricksimonian) [inline api/src/router/routes/github.js]: use `{}`  when if/else are on new lines (improves readability)

File: web/__tests__/Home.test.js [hunk 5]
```diff
@@ -79,27 +91,51 @@ describe('Home Component', () => {
...
-        <Home {...{ ...AUTH_PROPS_0, ...REQUEST_PROPS_0 }} />
+        <Home {...{ ...AUTH_PROPS_0, ...REQUEST_PROPS_1 }} />
+      </Router>
+    );
+
+    expect(getByTestId(TEST_IDS.APP.NEW_REQUEST)).toHaveTextContent('New Request');
+    expect(container.firstChild).toMatchSnapshot();
+  });
+
+  it('has link to create a new request for Requester when no open record exists', async () => {
```
Comment (patricksimonian) [inline web/__tests__/Home.test.js]: async func?

File: web/src/constants/form.js [hunk 6]
```diff
@@ -106,6 +120,140 @@ const requesterInfo = {
...
+    type: 'panel',
+    name: 'projectInfo',
+    title: '3. Project',
+    elements: [
+      {
+        type: 'text',
+        inputType: 'date',
+        name: 'prodDate',
+        dateFormat: 'yyyy-mm-dd',
+        title: 'Date of	release in production environment',
```
Comment (patricksimonian) [inline web/src/constants/form.js]: is there a reason for the big whitespace ? 'Date of _____'

File: web/src/constants/github.js [hunk 7]
```diff
@@ -70,7 +70,12 @@ export const getPrStatus = (state, merged, labels = []) => {
       // bceid rejected:
       else if (labels.includes(GITHUB_LABELS.BCEID_REJECTED)) return REQUEST_STATUS.REJECT;
       // when BCeID is requested, but not yet approved or completed, request is pending:
-      else if (labels.includes(GITHUB_LABELS.BCEID) && !labels.includes(GITHUB_LABELS.BCEID_APPROVED) && !labels.includes(GITHUB_LABELS.BCEID_COMPLETED)) return REQUEST_STATUS.PENDING;
+      else if (
```
Comment (patricksimonian) [inline web/src/constants/github.js]: use `{}`

File: web/src/containers/Home.js [hunk 8]
```diff
@@ -58,26 +65,29 @@ export class Home extends Component {
     }
 
     let requestsList = null;
+    let allowNewRequest = true;
+    // Show only error message when error loading the list:
     if (errorMessage) requestsList = errorMessage;
```
Comment (patricksimonian) [inline web/src/containers/Home.js]: use `{}`

File: web/src/containers/Home.js [hunk 9]
```diff
@@ -23,6 +27,9 @@ const StyledTableHeader = styled.div`
     color: white;
     background-color: #036;
   }
+  .hide {
+    display: none !important;
+  }
 `;
 
 export class Home extends Component {
```
Comment (patricksimonian) [inline web/src/containers/Home.js]: not a big deal but because this component only seems to be using one piece of state and _no_ lifecycle methods (componentDidMount etc). This component could be converted to a functional component and leverage `useState`

Comment (ShellyXueHan) [reply web/src/containers/Home.js]: thanks for pointing out, I might be using componentDidMount for this page soon cuz I just  realized that it's not refreshing when coming back to the page...

Comment (patricksimonian) [reply web/src/containers/Home.js]: useEffect maybe the best use of that since it will run on every re render. DidMount only runs the first time the component has mounted.

**Comment by ShellyXueHan** [comment]
that's outdated, new change is in this commit: https://github.com/bcgov/realm-o-matic/pull/61/commits/2d874fbe53540be9edf66fd222dbeac73ac6ceb1
